File: app/src/main/java/com/bopr/android/smailer/util/ResourceUtil.java
package com.bopr.android.smailer.util;

import android.content.Context;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.CharacterStyle;
import android.text.style.ForegroundColorSpan;
import android.text.style.ParagraphStyle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.bopr.android.smailer.PhoneEvent;
import com.bopr.android.smailer.R;
import com.bopr.android.smailer.util.draw.WavyUnderlineSpan;

import static androidx.core.content.ContextCompat.getColor;
import static com.bopr.android.smailer.PhoneEvent.STATE_IGNORED;
import static com.bopr.android.smailer.PhoneEvent.STATE_PENDING;
import static com.bopr.android.smailer.PhoneEvent.STATE_PROCESSED;

/**
 * Miscellaneous resource utilities.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public abstract class ResourceUtil {

    private ResourceUtil() {
    }

    public static int eventTypeImage(PhoneEvent event) {
        if (event.isSms()) {
            return R.drawable.ic_message;
        } else {
            return R.drawable.ic_call;
        }
    }

    public static int eventTypeText(PhoneEvent event) {
        if (event.isSms()) {
            if (event.isIncoming()) {
                return R.string.incoming_sms;
            } else {
                return R.string.outgoing_sms;
            }
        } else if (event.isMissed()) {
            return R.string.missed_call;
        } else {
            if (event.isIncoming()) {
                return R.string.incoming_call;
            } else {
                return R.string.outgoing_call;
            }
        }
    }

    public static int eventDirectionImage(PhoneEvent event) {
        if (event.isMissed()) {
            return R.drawable.ic_call_missed;
        } else if (event.isIncoming()) {
            return R.drawable.ic_call_in;
        } else {
            return R.drawable.ic_call_out;
        }
    }

    public static int eventStateImage(PhoneEvent event) {
        switch (event.getState()) {
            case STATE_PENDING:
                return R.drawable.ic_state_pending;
            case STATE_PROCESSED:
                return R.drawable.ic_state_done;
            case STATE_IGNORED:
                return R.drawable.ic_state_block;
        }
        return -1;
    }

    public static int eventStateText(PhoneEvent event) {
        switch (event.getState()) {
            case STATE_PENDING:
                return R.string.pending;
            case STATE_PROCESSED:
                return R.string.sent_email;
            case STATE_IGNORED:
                return R.string.ignored;
        }
        return -1;
    }

    /**
     * Returns text underlined with wavy red line.
     */
    public static Spannable underwivedText(Context context, String value) {
        Spannable spannable = new SpannableString(value);
        ParagraphStyle span = new WavyUnderlineSpan(context);
        spannable.setSpan(span, 0, spannable.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return spannable;
    }

    /**
     * Returns text of accent color.
     */
    public static Spannable accentedText(Context context, String value) {
        Spannable spannable = new SpannableString(value);
        CharacterStyle span = new ForegroundColorSpan(getColor(context, R.color.colorAccent));
        spannable.setSpan(span, 0, spannable.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return spannable;
    }

    public static void showToast(Context context, String text) {
        updateToastStyle(context, Toast.makeText(context, text, Toast.LENGTH_LONG)).show();
    }

    public static void showToast(Context context, int textRes) {
        updateToastStyle(context, Toast.makeText(context, textRes, Toast.LENGTH_LONG)).show();
    }

    private static Toast updateToastStyle(Context context, Toast toast) {
        View view = toast.getView();
        TextView textView = view.findViewById(android.R.id.message);
        view.setBackgroundResource(R.drawable.toast_frame);
        textView.setTextColor(getColor(context, R.color.toastForeground));
//        toast.setGravity(Gravity.CENTER, 0, 0);
        return toast;
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/util/ResourceUtil.kt
package com.bopr.android.smailer.util

import android.content.Context
import android.content.res.Resources
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import com.bopr.android.smailer.R
import com.bopr.android.smailer.messenger.telegram.TelegramException
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_BAD_RESPONSE
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_INVALID_TOKEN
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_NO_CHAT
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_NO_CONNECTION
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_NO_TOKEN
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_NO_UPDATES
import com.bopr.android.smailer.messenger.telegram.TelegramException.Code.TELEGRAM_REQUEST_FAILED
import com.bopr.android.smailer.messenger.MessageState
import com.bopr.android.smailer.messenger.MessageState.Companion.STATE_IGNORED
import com.bopr.android.smailer.messenger.MessageState.Companion.STATE_PENDING
import com.bopr.android.smailer.messenger.MessageState.Companion.STATE_PROCESSED
import com.bopr.android.smailer.provider.telephony.PhoneCallInfo
import java.util.Locale

/**
 * NOTE: To prevent drawables from being shrunk by R8's resource shrinker we have to
 * hold theirs hardcoded references.
 */
private val RES_DIRECTION_IMAGE = intArrayOf(
    R.drawable.ic_call_missed,
    R.drawable.ic_call_in,
    R.drawable.ic_call_out
)

/**
 * NOTE: To prevent drawables from being shrunk by R8's resource shrinker we have to
 * hold theirs hardcoded references.
 */
private val RES_STATE_IMAGE = intArrayOf(
    R.drawable.ic_hourglass,
    R.drawable.ic_state_done,
    R.drawable.ic_state_block
)

/**
 * NOTE: To prevent drawables from being shrunk by R8's resource shrinker we have to
 * hold theirs hardcoded references.
 */
private val RES_TYPE_IMAGE = intArrayOf(
    R.drawable.ic_message,
    R.drawable.ic_call
)

@DrawableRes
fun phoneCallTypeImage(info: PhoneCallInfo): Int {
    /* do not use direct drawable resources references here due to shrinker issue */
    return if (info.isSms) {
        RES_TYPE_IMAGE[0]
    } else {
        RES_TYPE_IMAGE[1]
    }
}

@DrawableRes
fun phoneCallDirectionImage(info: PhoneCallInfo): Int {
    /* do not use direct drawable resources references here due to shrinker issue */
    return when {
        info.isMissed ->
            RES_DIRECTION_IMAGE[0]

        info.isIncoming ->
            RES_DIRECTION_IMAGE[1]

        else ->
            RES_DIRECTION_IMAGE[2]
    }
}

@DrawableRes
fun messageStateImage(@MessageState state: Int): Int {
    /* do not use direct drawable resources references here due to shrinker issue */
    return when (state) {
        STATE_PENDING ->
            RES_STATE_IMAGE[0]

        STATE_PROCESSED ->
            RES_STATE_IMAGE[1]

        STATE_IGNORED ->
            RES_STATE_IMAGE[2]

        else ->
            throw IllegalArgumentException("Unknown state")
    }
}

@StringRes
fun phoneCallTypeText(info: PhoneCallInfo): Int {
    return if (info.isSms) {
        if (info.isIncoming) {
            R.string.incoming_sms
        } else {
            R.string.outgoing_sms
        }
    } else if (info.isMissed) {
        R.string.missed_call
    } else if (info.isIncoming) {
        R.string.incoming_call
    } else {
        R.string.outgoing_call
    }
}

@StringRes
fun phoneCallTypePrefix(info: PhoneCallInfo): Int {
    return if (info.isSms) {
        if (info.isIncoming) {
            R.string.incoming_sms_from
        } else {
            R.string.outgoing_sms_to
        }
    } else if (info.isMissed) {
        R.string.missed_call_from
    } else if (info.isIncoming) {
        R.string.incoming_call_from
    } else {
        R.string.outgoing_call_to
    }
}

@StringRes
fun messageStateText(@MessageState state: Int): Int {
    return when (state) {
        STATE_PENDING ->
            R.string.pending

        STATE_PROCESSED ->
            R.string.processed

        STATE_IGNORED ->
            R.string.ignored

        else ->
            throw IllegalArgumentException("Unknown state")
    }
}

@StringRes
fun onOffText(value: Boolean): Int {
    return if (value) R.string.on else R.string.off
}

@StringRes
fun telegramErrorText(error: TelegramException): Int {
    return when (error.code) {
        TELEGRAM_REQUEST_FAILED,
        TELEGRAM_BAD_RESPONSE -> R.string.error_sending_test_message

        TELEGRAM_NO_TOKEN -> R.string.no_telegram_bot_token

        TELEGRAM_INVALID_TOKEN -> R.string.bad_telegram_bot_token

        TELEGRAM_NO_UPDATES -> R.string.unable_determine_chat

        TELEGRAM_NO_CHAT -> R.string.require_start_chat

        TELEGRAM_NO_CONNECTION -> R.string.no_telegram_network_try_later
    }
}

fun Context.localeResources(locale: Locale): Resources {
    return if (locale == Locale.getDefault()) {
        resources
    } else {
        val configuration = resources.configuration.apply {
            setLocale(locale)
        }
        createConfigurationContext(configuration).resources
    }
}
